### PR TITLE
Events for plugins

### PR DESCRIPTION
**Pull request by italolelis**
## What does this PR do?
- Completely rewrites how plugins register hooks for events
- Added basic auth plugin
- Moved oauth2 resources to the plugin

### CODE REVIEW

File: cmd/janus/init.go [hunk 1]
```diff
@@ -18,6 +21,7 @@ var (
 	globalConfig *config.Specification
 	statsClient  stats.Client
```
Comment (rafaeljesus) [inline cmd/janus/init.go:22]: The variables below shouldn't be globally, otherwise we'll face concurrency issues

Comment (rafaeljesus) [reply cmd/janus/init.go:22]: Ah ok, I see now, you just declare them here and pass the way down to the dependencies on main.go

File: pkg/api/repository.go [hunk 2]
```diff
@@ -28,3 +37,29 @@ func exists(r Repository, def *Definition) (bool, error) {
 
 	return false, nil
 }
+
+// BuildRepository creates a repository instance that will depend on your given DSN
+func BuildRepository(dsn string, session *mgo.Session) (Repository, error) {
+	var repo Repository
+	dsnURL, err := url.Parse(dsn)
+	switch dsnURL.Scheme {
+	case "mongodb":
```
Comment (rafaeljesus) [inline pkg/api/repository.go]: perhaps
```go

const (
  mongodb = "mongodb"
  file = "file"
)

switch dsnURL.Scheme {
  case mongo:
  case file:
  default: 
}
```

File: pkg/api/repository.go [hunk 3]
```diff
@@ -28,3 +37,29 @@ func exists(r Repository, def *Definition) (bool, error) {
...
+		repo, err = NewMongoAppRepository(session)
+		if err != nil {
+			return nil, errors.Wrap(err, "Could not create a mongodb repository for api definitions")
+		}
+	case "file":
+		log.Debug("File system based configuration chosen")
+		var apiPath = dsnURL.Path + "/apis"
```
Comment (rafaeljesus) [inline pkg/api/repository.go]: `apiPath := fmt.Sprintf("%s/apis", dsnURL.Path)`

File: pkg/plugin/basic/handler.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+	"github.com/hellofresh/janus/pkg/errors"
+	"github.com/hellofresh/janus/pkg/opentracing"
+	"github.com/hellofresh/janus/pkg/render"
+	"github.com/hellofresh/janus/pkg/router"
+	log "github.com/sirupsen/logrus"
+)
+
+// Controller is the api rest controller
+type Controller struct {
```
Comment (rafaeljesus) [inline pkg/plugin/basic/handler.go]: Perhaps `Handler` is more common and understandable than `Controller`

File: pkg/plugin/basic/handler.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	repo Repository
+}
+
+// NewController creates a new instance of Controller
+func NewController(repo Repository) *Controller {
+	return &Controller{repo}
+}
+
+// Get is the find all handler
+func (c *Controller) Get() http.HandlerFunc {
```
Comment (rafaeljesus) [inline pkg/plugin/basic/handler.go]: Given this is a very basic crud handler, we could borrow some name conventions from some mcv frameworks out there, for this one we could rename to `Index()`

File: pkg/plugin/basic/handler.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+			errors.Handler(w, err)
+			return
+		}
+
+		render.JSON(w, http.StatusOK, data)
+	}
+}
+
+// GetBy is the find by handler
+func (c *Controller) GetBy() http.HandlerFunc {
```
Comment (rafaeljesus) [inline pkg/plugin/basic/handler.go]: This could be `Show()`

File: pkg/plugin/basic/handler.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+			errors.Handler(w, err)
+			return
+		}
+
+		render.JSON(w, http.StatusOK, data)
+	}
+}
+
+// PutBy is the update handler
+func (c *Controller) PutBy() http.HandlerFunc {
```
Comment (rafaeljesus) [inline pkg/plugin/basic/handler.go]: `Update()`

File: pkg/plugin/basic/handler.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+			errors.Handler(w, errors.New(http.StatusBadRequest, err.Error()))
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+// Post is the create handler
+func (c *Controller) Post() http.HandlerFunc {
```
Comment (rafaeljesus) [inline pkg/plugin/basic/handler.go]: Create()

File: pkg/plugin/basic/handler.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+			return
+		}
+
+		w.Header().Add("Location", fmt.Sprintf("/credentials/basic_auth/%s", user.Username))
+		w.WriteHeader(http.StatusCreated)
+	}
+}
+
+// DeleteBy is the delete handler
+func (c *Controller) DeleteBy() http.HandlerFunc {
```
Comment (rafaeljesus) [inline pkg/plugin/basic/handler.go]: `Destroy` or just `Delete`

File: pkg/plugin/basic/middleware.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+	return func(handler http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			log.Debug("Starting basic auth middleware")
+			logger := log.WithFields(log.Fields{
+				"path":   r.RequestURI,
+				"origin": r.RemoteAddr,
+			})
+
+			username, password, authOK := r.BasicAuth()
+			if authOK == false {
```
Comment (rafaeljesus) [inline pkg/plugin/basic/middleware.go]: `if !authOK`

File: pkg/plugin/basic/middleware.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+				errors.Handler(w, ErrNotAuthorized)
+				return
+			}
+
+			var found bool
+			users, err := repo.FindAll()
+			if err != nil {
+				log.WithError(err).Error("Error when getting all users")
+				errors.Handler(w, errors.New(http.StatusInternalServerError, "there was an error when lookin for users"))
+				return
```
Comment (rafaeljesus) [inline pkg/plugin/basic/middleware.go]: typo `looking`

File: pkg/plugin/basic/mongodb_repository.go [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	collectionName string = "basic_auth"
+)
+
+// User represents an user
+type User struct {
+	Username string `json:"username"`
```
Comment (rafaeljesus) [inline pkg/plugin/basic/mongodb_repository.go:15]: We should create uniq index for this field right?

File: pkg/plugin/oauth2/setup.go [hunk 13]
```diff
@@ -1,14 +1,27 @@
...
 	log "github.com/sirupsen/logrus"
 )
 
+var (
+	repo        Repository
+	adminRouter router.Router
+)
+
 func init() {
+	plugin.RegisterEventHook(plugin.StartupEvent, onStartup)
```
Comment (rafaeljesus) [inline pkg/plugin/oauth2/setup.go:28]: Not really sure yet if this is better, but maybe we could get rid of the globals above with the following design:
```go
oauth2Plugin := &OAuth2Plugin{repo: repo}
plugin.RegisterEventHook(plugin.StartupEvent, oauth2Plugin.onStartup)
plugin.RegisterEventHook(plugin.ReloadEvent, oauth2Plugin.onReload)
```

File: pkg/plugin/oauth2/setup.go [hunk 14]
```diff
@@ -19,14 +32,79 @@ type Config struct {
...
 }
 
-func setupOAuth2(route *proxy.Route, p plugin.Params) error {
+func onAdminAPIStartup(event interface{}) error {
+	e, ok := event.(plugin.OnAdminAPIStartup)
+	if !ok {
+		return errors.New("Could not convert event to admin startup type")
+	}
+
+	adminRouter = e.Router
```
Comment (rafaeljesus) [inline pkg/plugin/oauth2/setup.go:47]: and here we could have `p.adminRouter = e.Router` assigning the variable to the struct

File: pkg/plugin/oauth2/setup.go [hunk 15]
```diff
@@ -19,14 +32,79 @@ type Config struct {
...
+	}
+
+	switch dsnURL.Scheme {
+	case "mongodb":
+		repo, err = NewMongoRepository(e.MongoSession)
+		if err != nil {
+			return errors.Wrap(err, "Could not create a mongodb repository for oauth servers")
+		}
+	case "file":
+		var authPath = dsnURL.Path + "/auth"
```
Comment (rafaeljesus) [inline pkg/plugin/oauth2/setup.go]: `fmt.Sprintf("%s/auth", dsnURL.Path)`

File: pkg/plugin/oauth2/setup.go [hunk 16]
```diff
@@ -19,14 +32,79 @@ type Config struct {
...
+	}
+
+	config := e.Config.Database
+	dsnURL, err := url.Parse(config.DSN)
+	if err != nil {
+		return err
+	}
+
+	switch dsnURL.Scheme {
+	case "mongodb":
```
Comment (rafaeljesus) [inline pkg/plugin/oauth2/setup.go]: same on having constants

File: pkg/plugin/oauth2/setup.go [hunk 17]
```diff
@@ -48,11 +126,28 @@ func setupOAuth2(route *proxy.Route, p plugin.Params) error {
...
+	managerType, err := ParseType(oauthServer.TokenStrategy.Name)
 	if nil != err {
 		return nil, err
 	}
 
-	return oauth.NewManagerFactory(oauthServer).Build(managerType)
+	return NewManagerFactory(oauthServer).Build(managerType)
+}
+
+//loadOAuthEndpoints register api endpoints
```
Comment (rafaeljesus) [inline pkg/plugin/oauth2/setup.go]: give some space `// loadOAuthEndpoints`

File: pkg/plugin/plugin.go [hunk 18]
```diff
@@ -64,6 +57,48 @@ func RegisterPlugin(name string, plugin Plugin) error {
...
+// RegisterEventHook plugs in hook. All the hooks should register themselves
+// and they must have a name.
+func RegisterEventHook(name string, hook EventHook) error {
+	log.WithField("event_name", name).Debug("Event registered")
+
+	if name == "" {
+		return errors.New("event hook must have a name")
+	}
+
+	if hooks, dup := eventHooks[name]; dup {
```
Comment (rafaeljesus) [inline pkg/plugin/plugin.go:74]: I think we are missing lock here

File: pkg/plugin/plugin.go [hunk 19]
```diff
@@ -64,6 +57,48 @@ func RegisterPlugin(name string, plugin Plugin) error {
...
+	return nil
+}
+
+// EmitEvent executes the different hooks passing the EventType as an
+// argument. This is a blocking function. Hook developers should
+// use 'go' keyword if they don't want to block Caddy.
+func EmitEvent(name string, event interface{}) error {
+	log.WithField("event_name", name).Debug("Event triggered")
+
+	hooks, found := eventHooks[name]
```
Comment (rafaeljesus) [inline pkg/plugin/plugin.go:89]: missing lock here too

File: pkg/plugin/oauth2/setup.go (hunk 13 above)
Comment (italolelis) [reply pkg/plugin/oauth2/setup.go:28]: Why should we? 
They are encapsulated on the same package. Can you give me one advantage of this design so I can better understand it?

File: pkg/plugin/basic/mongodb_repository.go (hunk 12 above)
Comment (italolelis) [reply pkg/plugin/basic/mongodb_repository.go:15]: Yes probably

File: pkg/plugin/oauth2/setup.go (hunk 13 above)
Comment (rafaeljesus) [reply pkg/plugin/oauth2/setup.go:28]: That's for getting rid of the globals variables declared on the top, but perhaps we could keep an eyes on it, leave the way it is, and see if it really make sense to design the way I suggested

**Comment by rafaeljesus** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/185/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway) [Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/185/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway)

**Pull request merged**